pkg/controller/registry/resolver/resolver.go

```diff
@@ -183,11 +163,53 @@ func (r *SatResolver) SolveOperators(namespaces []string, csvs []*v1alpha1.Clust
 	return operators, nil
 }
 
-func (r *SatResolver) getSubscriptionInstallables(name, pkg, namespace string, current *Operator, catalog registry.CatalogKey, cachePredicates []OperatorPredicate, channelPredicates []OperatorPredicate, namespacedCache MultiCatalogOperatorFinder, visited map[OperatorSurface]*BundleInstallable) (map[solver.Identifier]solver.Installable, error) {
+func (r *SatResolver) getSubscriptionInstallables(sub *v1alpha1.Subscription, current *Operator, namespacedCache MultiCatalogOperatorFinder, visited map[OperatorSurface]*BundleInstallable) (map[solver.Identifier]solver.Installable, error) {
+	var cachePredicates, channelPredicates []OperatorPredicate
 	installables := make(map[solver.Identifier]solver.Installable, 0)
-	candidates := make([]*BundleInstallable, 0)
 
-	bundles := namespacedCache.Catalog(catalog).Find(cachePredicates...)
+	catalog := registry.CatalogKey{
+		Name:      sub.Spec.CatalogSource,
+		Namespace: sub.Spec.CatalogSourceNamespace,
+	}
+
+	var bundles []*Operator
+	{
+		var nall, npkg, nch, ncsv int
+
+		csvPredicate := True()
+		if current != nil {
+			// if we found an existing installed operator, we should filter the channel by operators that can replace it
+			channelPredicates = append(channelPredicates, Or(SkipRangeIncludes(*current.Version()), Replaces(current.Identifier())))
+		} else if sub.Spec.StartingCSV != "" {
+			// if no operator is installed and we have a startingCSV, filter for it
+			csvPredicate = WithCSVName(sub.Spec.StartingCSV)
+		}
+
+		cachePredicates = append(cachePredicates, And(
+			CountingPredicate(True(), &nall),
+			CountingPredicate(WithPackage(sub.Spec.Package), &npkg),
+			CountingPredicate(WithChannel(sub.Spec.Channel), &nch),
+			CountingPredicate(csvPredicate, &ncsv),
+		))
+		bundles = namespacedCache.Catalog(catalog).Find(cachePredicates...)
+
+		var si solver.Installable
+		switch {
+		case nall == 0:
+			si = NewInvalidSubscriptionInstallable(sub.GetName(), fmt.Sprintf("no operators found from catalog %s in namespace %s referenced by subscription %s", sub.Spec.CatalogSource, sub.Spec.CatalogSourceNamespace, sub.GetName()))
+		case npkg == 0:
```

Comment:
nit: is this actually `package %s not found in the catalog %s referenced by subscription %s`?

If we're saying the package was not found versus no operator (bundles) were found in that package. Because they are two different types of errors, empty package versus no package.

Reply:
Good observation -- we only know that a package (or channel) exists at this point if there is a bundle claiming to be in that package, so the wording is intentional.